src/NuGet.Clients/NuGet.Credentials/CredentialService.cs

```diff
@@ -0,0 +1,185 @@
+        /// Internal set is provided for use by unit tests.
+        /// </summary>
+        public IEnumerable<ICredentialProvider> Providers { get; internal set; }
+
+        /// <summary>
+        /// Provides credentials for http requests.
+        /// </summary>
+        /// <param name="uri">The uri of a web resource for which credentials are needed.</param>
+        /// <param name="proxy">The currently configured proxy. It may be necessary for CredentialProviders
+        /// to use this proxy in order to acquire credentials from their authentication source.</param>
+        /// <param name="isProxy">If true, get credentials to authenticate for the requested proxy.
+        /// If false, the credentials are intended for a remote service.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A credential object, or null if no credentials could be acquired.</returns>
+        public async Task<ICredentials> GetCredentials(Uri uri, IWebProxy proxy, bool isProxy,
+            CancellationToken cancellationToken)
+        {
+            if (uri == null)
+            {
+                throw new ArgumentNullException(nameof(uri));
+            }
+
+            ICredentials response = null;
+
+            foreach (var provider in Providers)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var retryKey = RetryCacheKey(uri, isProxy, provider);
+                var isRetry = _retryCache.ContainsKey(retryKey);
+
+                try
+                {
+                    // This local semaphore ensures one provider active per process.
+                    // We can consider other ways to allow more concurrency between providers, but need to
+                    // ensure that only one interactive dialogue is ever presented at a time, and that
+                    // providers are not writing shared resources.
+                    // Since this service is called only when cached credentials are not available to the caller,
+                    // such an optimization is likely not necessary.
+                    _providerSemaphore.WaitOne();
+
+                    if (!TryFromCredentialCache(uri, isProxy, isRetry, provider, out response))
+                    {
+                        response = await provider.Get(uri, proxy, isProxyRequest: isProxy, isRetry: isRetry,
+                            nonInteractive: _nonInteractive, cancellationToken: cancellationToken);
+                    }
+
+                    if (response != null)
+                    {
+                        AddToCredentialCache(uri, isProxy, provider, response);
+                        _retryCache[retryKey] = true;
+                        break;
+                    }
+                }
+                finally
+                {
+                    _providerSemaphore.Release();
+                }
+            }
+
+            return response;
+        }
+
```

Comment:
DefaultProviders are lazy loaded, and the delegate to initialized them are set up by the clients on startup.  Lazy load is so that we only report errors initializing them in cases where a command actually needs credentials.  One command reinitializes the credentialservice, so new instances will get the same defaults.



And